plugins/modules/files/filesize.py

```diff
@@ -0,0 +1,470 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2021, quidame <quidame@poivron.org>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = r'''
+---
+module: filesize
+
+short_description: Create a file with a given size, or resize it if it exists
+
+description:
+  - This module is a simple wrapper around C(dd) to create, extend or truncate
+    a file, given its size. It can be used to manage swap files (that require
+    contiguous blocks) or alternatively, huge sparse files.
+
+author:
+  - quidame (@quidame)
+
+version_added: "3.0.0"
+
+options:
+  path:
+    description:
+      - Path of the regular file to create or resize.
+    type: path
+    required: yes
+  size:
+    description:
+      - Requested size of the file.
+      - The value is a number (either C(int) or C(float)) optionally followed
+        by a multiplicative suffix, that can be one of C(B) (bytes), C(KB) or
+        C(kB) (= 1000B), C(MB) or C(mB) (= 1000kB), C(GB) or C(gB) (= 1000MB),
+        and so on for C(T), C(P), C(E), C(Z) and C(Y); or alternatively one of
+        C(K), C(k), C(KiB) or C(kiB) (= 1024B); C(M), C(m), C(MiB) or C(miB)
+        (= 1024KiB); C(G), C(g), C(GiB) or C(giB) (= 1024MiB); and so on.
+      - If the multiplicative suffix is not provided, the value is treated as
+        an integer number of blocks of I(blocksize) bytes each (float values
+        are rounded to the closest integer).
+      - When the I(size) value is equal to the current file size, does nothing.
+      - When the I(size) value is bigger than the current file size, bytes are
+        appended to the file without truncating it, in other words, without
+        modifying the existing bytes of the file.
+      - When the I(size) value is smaller than the current file size, it is
+        truncated to the requested value without modifying bytes before this
+        value.
+      - That means that a file of any arbitrary size can be grown to any other
+        arbitrary size, and then resized down to its initial size without
+        modifying its initial content.
+    type: str
```

Comment:
Why not simply use `type: bytes` and let Ansible handle the conversion? (That type uses `human_to_bytes` from Ansible's common module_utils.)

Reply:
oh... because it's not so simple :)

First, I didn't know that `type: bytes` was handled by a size converter, I really thought that this type was a python type, as in conversions like `b_path = to_bytes(path, errors='surrogate_or_strict')`. I'm surprised that `type='bytes'` in module's argument spec is related to size unit, since modules known to have a *size* option use either `str` (the most common) or `int` types, never `bytes` (afaik). OK, this is not a good argument (this is not because a function is not used that it is not usable).

And probably more important, I didn't want to reinvent the wheel (= to write a new `bytes_to_human()` function), but looking at the Ansible's common module_utils implementation, it doesn't define size units correctly, since `M` and `MB` are the same, and shouldn't: `KB`, `MB`, `GB` and the like should be SI compliant and are not (SI = powers of 10: 1km = 1 kilometer = 1000m, 1kg = 1 kilogram = 1000g, 1kB = 1 kilobyte = 1000B); Ansible's module_utils does not support `KiB`, `MiB`... (that are the same as `K`, `M`... powers of 2: 1KiB=1024B, 1MiB=1024KiB, ... and more explicitly different from SI's `KB`, `MB` than single letter units can be) and should. And *dd* does not handle bits, `b` is unit for *block* (POSIX) - I don't use it at all to avoid confusion.

That said, I'm not totally satisfied with `type='str'`, since the value may also be an integer or a float. When this is the case and the unitless value is not quoted (converted to string), user gets a warning. What about the `raw` type ? Is it usable here ?

Reply:
> oh... because it's not so simple :)

I was expecting that answer ;-)

> First, I didn't know that `type: bytes` was handled by a size converter, I really thought that this type was a python type, as in conversions like `b_path = to_bytes(path, errors='surrogate_or_strict')`. I'm surprised that `type='bytes'` in module's argument spec is related to size unit, since modules known to have a _size_ option use either `str` (the most common) or `int` types, never `bytes` (afaik). OK, this is not a good argument (this is not because a function is not used that it is not usable).

I only recently noticed that the `bytes` type exists, and then investigated to find out what it actually is. I was also expecting a Python `bytes`, though I knew that that won't really work since `bytes` cannot be safely transported through JSON (which is needed for modules). I would have named it differently I guess :)

> And probably more important, I didn't want to reinvent the wheel (= to write a new `bytes_to_human()` function), but looking at the Ansible's common module_utils implementation, it doesn't define size units correctly, since `M` and `MB` are the same, and shouldn't: `KB`, `MB`, `GB` and the like should be SI compliant and are not (SI = powers of 10: 1km = 1 kilometer = 1000m, 1kg = 1 kilogram = 1000g, 1kB = 1 kilobyte = 1000B); Ansible's module_utils does not support `KiB`, `MiB`... (that are the same as `K`, `M`... powers of 2: 1KiB=1024B, 1MiB=1024KiB, ... and more explicitly different from SI's `KB`, `MB` than single letter units can be) and should. And _dd_ does not handle bits, `b` is unit for _block_ (POSIX) - I don't use it at all to avoid confusion.

Welcome to the fun world of units ;-) The big advantage of using `bytes_to_human` is that it will be the same type of specification that's used "everywhere else" in Ansible. The big disadvantage is what you wrote above: that it has its own quirks (mostly caused by backwards compatibility I guess - changing its meaning of `MB` from mebibytes to SI would be a really tricks source of bugs)...

> That said, I'm not totally satisfied with `type='str'`, since the value may also be an integer or a float. When this is the case and the unitless value is not quoted (converted to string), user gets a warning. What about the `raw` type ? Is it usable here ?

`raw` would be possible, in that case you have to do all error handling yourself. Well, I guess it's not much worse than `str`, since for `str` everything is converted to a string, and you have to clean up the results yourself as well...

How about simplifying the list of suffixes a bit, so we only have `MB` (for SI) and `MiB` (for power-of-two), and not also `mB`, `m`, `M` and `miB`?

Reply:
Well, I choose the user-friendly way, that is: `type: raw`. It seems it doesn't need so much rework.

> How about simplifying the list of suffixes a bit, so we only have `MB` (for SI) and `MiB` (for power-of-two), and not also `mB`, `m`, `M` and `miB`?

Believe me or not, but all are supported by `dd` (at least mine). I would accept to drop `miB` family, it is so weird that I hope that nobody uses this format, but I'd rather keep the others, since they're far to be rare.

`kB` is the standard SI unit, `KB` is for compatibility; alternatively, `MB` is the standard SI unit, `mB` is for compatibility and is acceptable because there is no and there can't be *millibyte* (0.001 byte). Most `lvm` tools output sizes with lowercase single letters (`m`, `g`, `t`). And so on.

I think there are a lot of unpredictable cases of user input coming from arbitrary command output, so in that case, exhaustivity doesn't cost so much. IMO, things aren't so complicated: from developer point of view, anyone could find its way with this `SIZE_UNIT` dictionary. A little bit overkill, maybe, but very readable :); from end user point of view, it's really, really nice to not have to compute sizes and convert something here or there because we need uppercase here, lowercase there.

If it's really important for you, I can remove the `kiB`/`miB` family, but no more :)

Reply:
> Well, I choose the user-friendly way, that is: `type: raw`. It seems it doesn't need so much rework.

:+1:

> > How about simplifying the list of suffixes a bit, so we only have `MB` (for SI) and `MiB` (for power-of-two), and not also `mB`, `m`, `M` and `miB`?
> 
> Believe me or not, but all are supported by `dd` (at least mine). I would accept to drop `miB` family, it is so weird that I hope that nobody uses this format, but I'd rather keep the others, since they're far to be rare.

I totally believe you :)

The argument that most of these are output by other tools sounds good to me, so let's keep them. About the `miB` family, I would personally remove it, but I leave that decision to you.

> I think there are a lot of unpredictable cases of user input coming from arbitrary command output, so in that case, exhaustivity doesn't cost so much. IMO, things aren't so complicated: from developer point of view, anyone could find its way with this `SIZE_UNIT` dictionary. A little bit overkill, maybe, but very readable :); from end user point of view, it's really, really nice to not have to compute sizes and convert something here or there because we need uppercase here, lowercase there.

True. The main disadvantage for users is that this large amount of aliases for a unit is that it makes the documentation more complex, which decreases readability. It's not that bad that it definitely must be improved, though :)

Reply:
`miB` suffix family: dropped